### PR TITLE
Switch QTT and STT over to using Pojos to deserialize Json AND add keyField tests

### PR DESCRIPTION
**Pull request by big-andy-coates**
# UPDATED Description 

This PR contains two main changes (combined into one PR as the initial PR was still not reviewed by the time the next set of changes was ready):
-  I want to extend the QTTs. Before doing some, I've switched them over to use Java Pojos for handling the deserialization of the test files, rather than manually manipulating Json nodes.
- Next, I've extended QTT to support testing the sources, (i.e. streams and tables), in the metastore _after_ the statements have been run. The main reason for doing this was to allow me to add tests around the state of the `keyField` in the sources.

There are inconsistencies around the handling of the keyField of sources added to the metastore.  This has implications for downstream queries.  I'd like to clean up these inconsistencies _before_ adding structured keys. However, such changes may have backwards compatibility issues. Hence...

# ... the main thing I'm looking for from this PR is people's careful attention on the implications of fixing the issues I've highlighted below.

## Issues found with current implementation around `keyField`.

The tests added assert the key fields current value, even where this is _wrong_.   I've done this to highlight the current issues and to ensure that any changes made, as part of the structured keys work, can not inadvertently fix any of these issues without a test failing. This will force us to think about the backwards compatibility of such fixes.

For test cases where the key field is currently set incorrectly, I've added an `issues` comment to the test to track/highlight the issue. I've not added issues where the table keyfield is `null`, as we're dropping that requirement anyway.

## Summary of issues:

The issues look to fall into one of several categories, (more detailed explanation below):
1. Setting the name of the key field incorrectly: fixing this could result in a repartition step being dropped, which would break backwards compatibility.
2. Setting the type of the key field incorrectly: fixing this could result in existing joins being rejected due to different key types, which would break backwards compatibility.
3. Setting a key field when none should be set: I can't think of any compatibility this would throw up... can you?

## Considerations to fixing these:

Fixing these issues isn't straight forward due to the fact that it's not the processing of the statement itself that is effected, but all downstream queries.  This is because the statement itself runs just fine, but it pollutes the metastore with bad data.

Ideally, we'd clean up these issues before layering structured key support on top. (Building on solid foundations, etc). However, there are backwards compatibility concerns. Maintaining backwards compatibility isn't going to be pretty in the code. It will likely involve passing both a `legacyKeyField` and a new `keyField`, and choosing between them via config. Anything else is likely going to miss edge cases or implications we've not considered.  We would have the _option_ of removing the `legacyKeyField` in the next major version bump to clean things up a little, (though this would break some existing queries, so is a tricky call).

Regarding the config setting we use to switch between legacy and fixed key fields... I'm not really a fan of having multitudes of CompatibilityBreakingConfig that should never really be set by a user.  What do people think to the idea of adding a _ksql_version_ config?  We could have this automatically be set from the version of KSQL you're running and stored in the properties of each command in the command topic.  Then, rather than a multitude of configs, we can just have one and switches in the code for different code paths depending on the version of KSQL the query was started with.  

Thoughts on the above considerations / suggestions much appreciated!

#### Implications
*Current Implications*: Where the current implementation is incorrectly setting the name of the key field, any downstream query may incorrectly add a repartition step where none is needed.

*Implications if we fix*:  If we fix this, then downstream queries that _previously_ added a repartition step will stop doing so after upgrading.  *OUTSTANDING QUESTION*: is this a backwards compatible change? I'm assuming so, as we could have messages in the repartition topic that haven't been processed by later steps. Without complex 'drain' logic during the upgrade I'm assuming we'll need a compatibility breaking config to 'fix' this. 

Examples of when the key field name is set incorrect are:

#### Projections that alias the key

e.g.

```
-- given:
CREATE STREAM INPUT (foo INT) WITH (key='foo', ...);
-- when:
CREATE STREAM OUTPUT AS SELECT foo as aliased FROM INPUT;
```

Will result in the key field of `OUTPUT` in the metastore having the name `FOO`, where as it should pick up the alias and be `ALIASED`.

#### GROUP BY where the new key field has an internal name

e.g. 

```
CREATE TABLE OUTPUT AS SELECT foo, COUNT(*) FROM INPUT GROUP BY foo;
```

Will result in the key field of `OUTPUT` in the metastore having the name `KSQL_INTERNAL_COL_0`, where as it should be `FOO`. 

#### Implications
*Current implications*: Users may get an error when trying to perform a join between two compatible streams, or be allowed to join two incompatible streams.

*Implications of fixing*: None, where a join was previously incorrectly not allowed. Where a join was previously allowed on incompatible streams the fix would break backwards compatibility. Note: though the streams may be 'incompatible' that does not mean the join would not work, e.g. joining streams with INT and BIGINT keys would work just fine for many key values. Likewise, a STRING key can contain an integer, and join just fine.

Examples of when the key field type is set incorrect are:

#### GROUP BY statements that change the key

e.g.

```
-- given:
CREATE STREAM INPUT (foo INT, bar INT) WITH (kafka_topic='input_topic', value_format='JSON');
-- when:
CREATE TABLE OUTPUT AS SELECT foo AS Aliased, COUNT(*) FROM INPUT GROUP BY foo;
```

Will result in the OUTPUT table in the metastore incorrectly having the `keyField`'s schema being set to `STRING`, where as it should be `INT`.

Note: this may initially seem correct - after all the key _is_ a STRING in the current implementation.  Unfortunately, this is inconsistent with other uses of `GROUP BY`, where the key is not changed, or any use of `PARTITION BY`, both of which would result in an `INT` keyfield type.

#### Implications
*Current Implications*:  All found instances result in setting the key field to a name that does not exist in the value schema.  I _believe_ this is benign. It _may_ be possible that a re-partition does NOT occur when one is needed if the keyField ends up being set to a field name that a downstream query tries to repartition by, however I've not tested if this is the case.

*Implications if we fix*:  Likely nothing.  If a re-partition is not occurring in a downstream query where one should be, then the downstream query is broken and hence fixing this should not break backwards compatibility.

Examples of when the key field is set incorrect are:

#### Queries that drop the key field

e.g. 

```
-- given
CREATE STREAM INPUT (foo INT, bar INT) WITH (key='foo', ...);
-- when:
CREATE STREAM OUTPUT AS SELECT bar FROM INPUT;
```

Will result in the OUTPUT stream in the metastore incorrectly having the `keyField` still set to `FOO`, where as it should be `null`.

### Testing done 
no production code changed.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by agavra** [comment]
> Implications if we fix: If we fix this, then downstream queries that previously added a repartition step will stop doing so after upgrading. OUTSTANDING QUESTION: is this a backwards compatible change?

If I'm doing an aggregation of some sort, and then I lose the repartition topic, is it possible that my aggregation is messed up somehow? (e.g. when I roll out the new version, there will be a period of time when I can't look back far enough in the new topic or something like that)

File: ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java [hunk 1]
```diff
@@ -493,6 +502,7 @@ public JsonNode getNode() {
   }
 
   static class TestCase {
+
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java]: That's our coding standard...

File: ksql-engine/src/test/resources/query-validation-tests/key-field.json [hunk 2]
```diff
@@ -0,0 +1,697 @@
+{
+  "comments": [
+    "Tests covering handling of key-fields.",
+    "The 'key-field' is an optional field within the value that matches the key",
+    "",
+    "There are four main dimensions to this test suite:",
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/key-field.json]: lol

File: ksql-engine/src/test/resources/query-validation-tests/key-field.json [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+          {"name": "INPUT", "type": "stream", "keyField": null},
+          {"name": "OUTPUT", "type": "table", "keyField": {"name": "KSQL_INTERNAL_COL_0", "schema": {"type": "STRING"}}}
+        ]
+      }
+    },
+    {
+      "name": "stream | initially null | group by (-) | key not in value | -",
+      "statements": [
+        "CREATE STREAM INPUT (foo INT, bar INT) WITH (kafka_topic='input_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM INPUT GROUP BY foo;"
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/key-field.json]: This is just current functionality.  No production code has changed.

File: ksql-engine/src/test/resources/query-validation-tests/key-field.json [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+      ],
+      "inputs": [
+        {"topic": "INTERMEDIATE", "key": "x", "value": {"bar": 1}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": "x", "value": {"BAR": 1}}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "table", "keyField": null}
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/key-field.json]: See my comment below.

**Comment by big-andy-coates** [comment]
Thanks for the review @hjafarpour. However, I think you may have missed the point of this PR.  Did you read the description at the top?

This PR adds tests covering _current functionality_ of KSQL. There is no functional change.  

What I'm mostly looking for from this PR is thoughts on the implications of fixing each of the issues I've highlighted in the description.

**Comment by hjafarpour** [comment]
As we discussed offline, I'm inclined to break the backward compatibility on this case and not support the current key setting logic for the existing queries. 
As you also mentioned there are cases where the key is set incorrectly and this may result in unnecessary repartitioning in the downstream queries. This is results in performance degradation but does not affect correctness, however, we should also check to see if the incorrect key setting would result in omission of a necessary reapportioning in the downstream queries which would affect the correctness.
In summary, I suggest we ignore the current implementation when we implement the structured key support and start with a clean design!

**Comment by big-andy-coates** [comment]
FYI, I've now committed this in the 5.2 branch, so this PR now has virtually no changes.
